### ISSUE TITLE
Encoding information is incomplete

### ISSUE THREAD

**Issue by DarrenHoffman** [MicrosoftDocs/typography-issues#814]
In some cases multiple characters are represented by a single glyph-form, but the encoding info only covers single character glyph forms. Clearly Microsoft Word is aware of the mappings so they definitely exist. It would be good to add them to the documentation.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: f3625191-cbf0-c30b-2f89-60ed8c11d4d7
* Version Independent ID: 7e7a7e77-9514-fcc9-82e8-36335e0f222e
* Content: [Legacy Arabic Font Encodings - Typography](https://docs.microsoft.com/en-us/typography/legacy/legacy_arabic_fonts)
* Content Source: [typographydocs/legacy/legacy_arabic_fonts.md](https://github.com/MicrosoftDocs/typography/blob/live/typographydocs/legacy/legacy_arabic_fonts.md)
* Product: **typography**
* GitHub Login: @PeterCon
* Microsoft Alias: **alib**

**Comment by PeterConstable** [comment]
>The difference in appearance is stark, though an Arabic Language reader may recognize these two forms as equivalent as far as I know (I myself do not know how to read Arabic at all), but surely the single glyph form is preferable if not absolutely required.

The single glyph form is absolutely required.

The intent of the documentation was not to provide all information needed to support Arabic script, but only to document the legacy 8-bit encodings* associated with ARABIC_CHARSET_SIMPLIFIED and ARABIC_CHARSET_TRADITIONAL. I think the article does that.

\* Implementations using the legacy symbol encoding would store strings as 8-bit characters, but then in the display pipleline map to 16-bit code points in the 0xFxxx range (which happens to correspond to Unicode PUA characters). This is mentioned in that article (see [The nature of the font encodings](https://docs.microsoft.com/en-us/typography/legacy/legacy_arabic_fonts#the-nature-of-the-font-encodings)).

**Comment by DarrenHoffman** [comment]
I'm not asking for help with general Arabic encoding. I am specifically working with legacy fonts that only have a 3,0 cmap table and where the fSelection upper byte code is ARABIC_CHARSET_TRADITIONAL. 

And, I am still not convinced the legacy 8-bit encodings are fully documented here. I have a set of these legacy fonts in hand at the moment. They contain almost 256 glyphs mapped to values in the 0xF200 to 0xF2FF range of the 3,0 cmap table. So that's essentially an 8-bit encoding. There is absolutely no other encoding info in these fonts. 

And yet Microsoft Word is able to map the pair of Unicode values I mentioned above to one of the presentation forms in the glyph that is not included in the mappings provided by this web page. All the mappings provided by this web page are consistent with what appears in the font files I have. 
So this certainly implies Microsoft Word has some further information about the encodings in these fonts that is not provided in this web page.

**Comment by DarrenHoffman** [comment]
So maybe these 3 screen shots can do a better job of illustrating my issue than all that text in the previous post:
The font I am using, Sultan Normal, it is a ARABIC_CHARSET_TRADITIONAL font, the only Unicode character codes encoded are the PUA codes 0xF200 - 0xF2FF. (in a 3,0 cmap).
You can see the glyph displayed in the Word file is an exact match for the glyph at 0xF2DC in the font which can be seen in the FontForge screenshot.
You can see the underlying Unicode text in the DOCX file itself is actually <U+644 U+627> in the emacs window.
So, HOW does Word know to use the glyph at 0xF2DC?! That one isn't included in the mapping on the ARABIC_CHARSET_TRADITIONAL's documentation here. 
![image](https://user-images.githubusercontent.com/33430105/128963660-80962437-1c64-4834-be2f-9d9bae0c32d4.png)
![image](https://user-images.githubusercontent.com/33430105/128963693-287f89f9-ad92-4b8c-a1a3-0940669b1581.png)
![image](https://user-images.githubusercontent.com/33430105/128963736-3b21d597-2f75-4f6b-9489-d9931a0127cb.png)

**Comment by DarrenHoffman** [comment]
Do you happen to know offhand how many of these orthographically required ligating forms are used with ARABIC_CHARSET_TRADITIONAL. I can't personally tell just by eyeballing the glyphs in the font file, though I guess I could probably figure it out sooner or later with a close examination.

**Comment by khaledhosny** [comment]
The documentation indeed does not cover any ligatures. I had to reverse-engineer them by looking at the fonts and double-checking my guess with what Uniscribe does. The ligature table I came up with is [here](https://github.com/harfbuzz/harfbuzz/blob/74a5a8d9c17c88e521dd7f10211d14f57dc6ac94/src/hb-ot-shape-complex-arabic-pua.hh) (part of https://github.com/harfbuzz/harfbuzz/pull/3063).

**Comment by PeterCon** [comment]
A number of glyphs are presentation ligature forms for combinations other than orthographically-required ligatures. In your screen shot above, F2D6 and F2D7 are initial and isolate forms for the combination DAL + HAMZA ABOVE (in Unicode, the sequence <U+062F, U+0654>.

**Comment by PeterCon** [comment]
@khaledhosny In your harfbuzz data, you have F21D, F21E and F21F (for Traditional) as 
* BEH WITH YEH **ISOLATED** FORM
* TEH WITH YEH **ISOLATED** FORM
* NOON WITH YEH **ISOLATED** FORM

In the Royal Arabic font, though, these are final forms. 

In Microsoft's legacy Arabic engine (25-year-old code) I can't easily tell which ligatures are assumed to be initial, medial final or isolated (there's one array that combines all ordered by component count then legacy code point), but the comments for F212..F214 and F21C..F21F indicate those are final. 

For ligatures beginning with LAM, I'm not familiar enough with Arabic script to tell in the Royal Arabic font whether these are isolate, final or intended to be used as both.

![image](https://user-images.githubusercontent.com/12788502/129122430-dfac4d7e-0134-4b53-a627-fc042fc53e97.png)

Royal Arabic is just one representative of TRADITIONAL that I happen to have looked at. I have no way of knowing what to consider truth—a sample font or source code, either of which may be buggy. (I think I've found one bug in the array shown in the article: for 0638 isolate, I suspect 0xa3 s/be 0xa6.)

**Comment by DarrenHoffman** [comment]
> The documentation indeed does not cover any ligatures. I had to reverse-engineer them by looking at the fonts and double-checking my guess with what Uniscribe does. The ligature table I came up with is [here](https://github.com/harfbuzz/harfbuzz/blob/74a5a8d9c17c88e521dd7f10211d14f57dc6ac94/src/hb-ot-shape-complex-arabic-pua.hh) (part of [harfbuzz/harfbuzz#3063](https://github.com/harfbuzz/harfbuzz/pull/3063)).

thank you very much. That will probably be very helpful.

**Comment by DarrenHoffman** [comment]
> Royal Arabic is just one representative of TRADITIONAL that I happen to have looked at. I have no way of knowing what to consider truth—a sample font or source code, either of which may be buggy. (I think I've found one bug in the array shown in the article: for 0638 isolate, I suspect 0xa3 s/be 0xa6.)

Yes, I am sure you are correct about that bug. 0xa6 is clearly the isolate. And I don't believe any of the characters should have 3 distinct glyph forms as would be the case if 0xa3 were correct. You either have 1 form (essentially to isolate for all cases, but that's not common), or 2 forms (one for Isolate+Initial and one for Medial+Final), or all 4 distinct forms. Any glyph that has at least 3 distinct glyphs would also have the 4th form distinct as well, it wouldn't make sense not to, I'm pretty sure). 
And you can see 0xa3 has the trailing connector stroke that would make it a distinct Initial form.

**Comment by PeterConstable** [comment]
@DarrenHoffman Do you know of fonts that use SIMPLIFIED charset?

**Comment by DarrenHoffman** [comment]
> @DarrenHoffman Do you know of fonts that use SIMPLIFIED charset?

no, I just have this one set of TRADITIONAL font faces. I can check with my customer but I think it's just this one family they are using.

**Comment by ebraminio** [comment]
> @DarrenHoffman Do you know of fonts that use SIMPLIFIED charset?

There are just lots of such fonts here, http://www.aboaziz.net/misc/arabic%20fonts%20pack/ and [1](http://www.aboaziz.net/misc/arabic%20fonts%20pack/%CE%D8%E6%D8%20%DA%D1%C8%ED%C9%20%E1%E1%CA%CD%E3%ED%E1/akek%23b%23.TTF) and [2](http://www.aboaziz.net/misc/arabic%20fonts%20pack/%CE%D8%E6%D8%20%DA%D1%C8%ED%C9%20%E1%E1%CA%CD%E3%ED%E1/akek%23n%23.TTF) and [3](http://www.aboaziz.net/misc/arabic%20fonts%20pack/%CE%D8%E6%D8%20%DA%D1%C8%ED%C9%20%E1%E1%CA%CD%E3%ED%E1/akek%23xb%23.TTF) are simplified ones apparently. There are also some others like [4](http://www.aboaziz.net/misc/arabic%20fonts%20pack/%CE%D8%E6%D8%20%DA%D1%C8%ED%C9%20%E3%CC%E3%DA%E5/LUCASIT.TTF) which apparently are different with rest, which don't have 0xB3 (traditional) and 0xB2 (simplified) and have 0xEE which apparently we don't know what it means.

**Comment by DarrenHoffman** [comment]
@ebraminio  ah, thank you very much.

**Comment by ebraminio** [comment]
> For ligatures beginning with LAM, I'm not familiar enough with Arabic script to tell in the Royal Arabic font whether these are isolate, final or intended to be used as both.

If that helps on understanding the code around better,

These are `init`:
<img width="261" alt="image" src="https://user-images.githubusercontent.com/833473/129282011-eada5ba0-90b3-4474-a32b-66f754724152.png">
<img width="258" alt="image" src="https://user-images.githubusercontent.com/833473/129282202-a4f72611-a170-4e97-8330-bb7ba485d8e9.png">

These should be `isol`:
<img width="386" alt="image" src="https://user-images.githubusercontent.com/833473/129282159-4fb01cd9-8913-4817-ab48-0572fa54e318.png">
<img width="134" alt="image" src="https://user-images.githubusercontent.com/833473/129282225-c294d890-6c7e-484c-9d1e-fd120c1617dd.png">

These should be `fina`:
<img width="137" alt="image" src="https://user-images.githubusercontent.com/833473/129282281-44ba62f5-d2e0-4e91-84ca-781a65feca5f.png">
<img width="387" alt="image" src="https://user-images.githubusercontent.com/833473/129282292-ed4a74cd-ca36-45aa-8ba3-b80d9df03ee4.png">

**Comment by khaledhosny** [comment]
> @khaledhosny In your harfbuzz data, you have F21D, F21E and F21F (for Traditional) as

You are right, I probably was looking into a font that didn’t differentiate enough between final and isolated forms of these ligatures (the table was based on some python tool I wrote long ago and I don’t remember all the details). I double checked by shaping with Uniscribe and the ligatures were indeed used in final form (i.e. between medial and final glyphs).

**Comment by DarrenHoffman** [comment]
> It would be nice to know, for example, where in the display chain the ligature mapping is done: is it at the top level of Arabic character input, or after the joining form mapping?

well I noticed that in the font I have there is a ligature for I believe "Allah", except that the ligature's glyph in the font file doesn't actually include the first character, the alif (I think that's the one) character. But the only Unicode representation I can find for the ligature includes the Alif. So with that in mind, I guess when using these fonts it might be better to map Unicode characters to the font's PUA codes before performing ligature substitution since there doesn't seem to be a clean Unicode mapping. But I'm not sure, that's just one thing I noticed. But it seems to require a slightly different ligature mapping rules than for a modern Unicode font.

**Comment by khaledhosny** [comment]
There at least three ligatures with no corresponding Unicode Presentation Forms code points, 0xF201 as you noted, as well as 0xF211 and 0xF2EE.

**Comment by PeterCon** [comment]
@DarrenHoffman @khaledhosny I've updated the page with complete details on the presentation encoding for Traditional Arabic. Please review—in particular, let me know if the presentation works.

At some point, if some representative Simplified fonts can be obtained, a similar update might be done for that.

**Comment by khaledhosny** [comment]
Looks good, but for entries like:
> 064B   fathatan 	F2E7 or F2F5

I understand these are the high and low variants, but there is no explanation how they are used. I have always assumed this provides `mset`-like behavior, but stating this explicitly (if it is really the case) might be a good idea.

**Comment by PeterCon** [comment]
@khaledhosny I've updated per your comments.

**Comment by khaledhosny** [comment]
> @khaledhosny I've updated per your comments.

Looks good, thanks.

**Comment by moi15moi** [comment]
It seems that there are also hebrew symbol font. Ex: [Dor.zip](https://github.com/user-attachments/files/31046858/Dor.zip)

It would be nice if the mapping could be documented like for arabic font: https://learn.microsoft.com/en-us/typography/legacy/legacy_arabic_fonts

It would help harfbuzz supporting these type of fonts (https://github.com/harfbuzz/harfbuzz/issues/6159)
